File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/security/filter/JWTAuthenticationFilter.java
package com.ntnu.gidd.security.filter;

import com.ntnu.gidd.security.UserDetailsImpl;
import com.ntnu.gidd.security.config.JWTConfig;
import com.ntnu.gidd.util.JwtUtil;
import io.jsonwebtoken.ExpiredJwtException;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.ArrayList;
import java.util.UUID;

@Slf4j
@AllArgsConstructor
public class JWTAuthenticationFilter extends OncePerRequestFilter {

    private JWTConfig jwtConfig;
    private JwtUtil jwtUtil;

    /**
     * Read authentication from request header and add to security context.
     */
    @Override
    protected void doFilterInternal(HttpServletRequest request,
                                    HttpServletResponse response,
                                    FilterChain chain) throws IOException, ServletException {
        String header = extractAuthorizationHeaderFromRequest(request);
        log.debug("Filtering Jwt from request header: {}", header);

        if (header == null || !header.startsWith(jwtConfig.getPrefix())) {
            chain.doFilter(request, response);
            return;
        }

        processAuthentication(request);
        chain.doFilter(request, response);
    }

    private String extractAuthorizationHeaderFromRequest(HttpServletRequest request) {
        return request.getHeader(jwtConfig.getHeader());
    }

    /**
     * Read the JWT from Authorization header and validate the token.
     */
    private void processAuthentication(HttpServletRequest request) {
        String token = extractAuthorizationHeaderFromRequest(request);

        try {
            setAuthenticationFromToken(token);
        } catch (ExpiredJwtException ex) {
            log.error("JWT expired", ex);
            request.setAttribute("exception", ex);
        } catch (BadCredentialsException ex) {
            log.error("Error while authenticating user:", ex);
            request.setAttribute("exception", ex);
        } catch (Exception ex) {
            log.error("Error while authenticating user, clearing security context: ", ex);
            SecurityContextHolder.clearContext();
        }
    }

    private void setAuthenticationFromToken(String token) {
        if (tokenIsValid(token)) {
            setAuthentication(token);
        }
    }

    private boolean tokenIsValid(String token) {
        return token != null && jwtUtil.isValidToken(token);
    }

    private void setAuthentication(String token) {
        String email = jwtUtil.getEmailFromToken(token);
        UUID id = jwtUtil.getUserIdFromToken(token);
        UserDetails userDetails = UserDetailsImpl.builder()
                .id(id)
                .email(email)
                .build();

        UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(
                userDetails, null, new ArrayList<>());

        log.info("Authenticating user: {}", authentication);
        SecurityContextHolder.getContext().setAuthentication(authentication);
    }

}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/filter/JWTAuthenticationFilter.kt
package ntnu.idatt2105.security.filter

import io.jsonwebtoken.ExpiredJwtException
import ntnu.idatt2105.security.JwtUtil
import ntnu.idatt2105.security.config.JWTConfig
import ntnu.idatt2105.user.service.UserDetailsImpl
import org.springframework.http.HttpStatus
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.web.filter.OncePerRequestFilter
import org.springframework.web.server.ResponseStatusException
import java.io.IOException
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JWTAuthenticationFilter(val jwtConfig: JWTConfig, val jwtUtil: JwtUtil) : OncePerRequestFilter() {

    @Throws(IOException::class, ServletException::class)
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {
        val header = extractAuthorizationHeaderFromRequest(request)
        if (header != null && header.startsWith(jwtConfig.prefix)) {
            processAuthentication(request)
            chain.doFilter(request, response)
            return
        }
        chain.doFilter(request, response)
        return
    }

    private fun extractAuthorizationHeaderFromRequest(request: HttpServletRequest): String? {
        return request.getHeader(jwtConfig.header)
    }

    private fun processAuthentication(request: HttpServletRequest) {
        try {
            val token = extractAuthorizationHeaderFromRequest(request) ?: throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
            setAuthenticationFromToken(token)
        } catch (ex: ExpiredJwtException) {
            request.setAttribute("exception", ex)
        } catch (ex: BadCredentialsException) {
            request.setAttribute("exception", ex)
        } catch (ex: Exception) {
            SecurityContextHolder.clearContext()
        }
    }

    private fun setAuthenticationFromToken(token: String) {
        if (tokenIsValid(token)) {
            setAuthentication(token)
        }
    }

    private fun tokenIsValid(token: String?): Boolean {
        return token != null && jwtUtil.isValidToken(token)
    }

    private fun setAuthentication(token: String) {
        val email: String? = jwtUtil.getEmailFromToken(token)
        val id: UUID? = jwtUtil.getUserIdFromToken(token)
        val roles: List<SimpleGrantedAuthority> = jwtUtil.getRolesFromToken(token)
        val userDetails: UserDetails = UserDetailsImpl(id!!, email!!, "", setOf())
        val authentication = UsernamePasswordAuthenticationToken(
                userDetails, null, roles)
        SecurityContextHolder.getContext().authentication = authentication
    }
}
